serviceregistration/kubernetes/retry_handler.go

```diff
@@ -22,58 +25,86 @@ type retryHandler struct {
 	// To synchronize setInitialState and patchesToRetry.
 	lock sync.Mutex
 
-	// setInitialState will be nil if this has been done successfully.
-	// It must be done before any patches are retried.
-	setInitialState func() error
+	// initialStateSet determines whether an initial state has been set
+	// successfully or whether a state already exists.
+	initialStateSet bool
+
+	// State stores an initial state to be set
+	initialState sr.State
 
 	// The map holds the path to the label being updated. It will only either
 	// not hold a particular label, or hold _the last_ state we were aware of.
 	// These should only be updated after initial state has been set.
 	patchesToRetry map[string]*client.Patch
-}
 
-func (r *retryHandler) SetInitialState(setInitialState func() error) {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	if err := setInitialState(); err != nil {
-		if r.logger.IsWarn() {
-			r.logger.Warn(fmt.Sprintf("unable to set initial state due to %s, will retry", err.Error()))
-		}
-		r.setInitialState = setInitialState
-	}
+	// client is the Client to use when making API calls against kubernetes
+	client *client.Client
 }
 
 // Run must be called for retries to be started.
-func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup, c *client.Client) {
+func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	wait.Add(1)
+
+	r.setInitialState(shutdownCh)
+
 	// Run this in a go func so this call doesn't block.
 	go func() {
 		// Make sure Vault will give us time to finish up here.
-		wait.Add(1)
 		defer wait.Done()
 
-		retry := time.NewTicker(retryFreq)
-		defer retry.Stop()
-		for {
-			select {
-			case <-shutdownCh:
-				return
-			case <-retry.C:
-				r.retry(c)
+		var g run.Group
+
+		// This run group watches for the shutdownCh
+		g.Add(func() error {
+			<-shutdownCh
+			return nil
+		}, func(error) {})
+
+		checkUpdateStateStop := make(chan struct{})
+		g.Add(func() error {
+			r.periodicUpdateState(checkUpdateStateStop)
+			return nil
+		}, func(error) {
+			close(checkUpdateStateStop)
+			r.client.Shutdown()
+		})
+
+		g.Run()
+	}()
```

Comment:
Question, what were the main goals with refactoring the retry handler? Were they mainly to make sure it's thread safe?

Reply:
That's correct. Even though this logic is running within a goroutine, `r.retry(c)` would block until the API request finishes. If you were to interrupt via `^C`, it would hang until the next time the the for loop runs, which is blocked by `c.retry(c)`.

Reply:
Gotcha. Thanks, that helps!